microscoped-core/src/main/java/org/tomitribe/microscoped/core/Scope.java

```diff
@@ -53,29 +76,60 @@ public <T> T get(final Contextual<T> contextual, final CreationalContext<T> crea
      * @return existing the bean instance or null
      */
     public <T> T get(final Contextual<T> contextual) {
-        return (T) instances.getOrDefault(contextual, NOTHING).get();
+        final CompletableFuture<Instance> future = instances.get(contextual);
+        return fastWaitForValue(future);
+    }
+
+    private static <T> T fastWaitForValue(final CompletableFuture<Instance> future) throws Error {
+        final T value;
+        if (future == null) {
+            value = null;
+        } else if (future.isDone() && !future.isCompletedExceptionally()) {
+            // Completed normally: read the result without join()
+            final Instance<T> instance = future.getNow(null); // never null in this branch
+            value = instance.get();
+        } else {
+            try {
+                final Instance<T> instance = future.join();
+                value = instance.get();
+            } catch (final CompletionException ce) {
+                // Try to hide internal semantics here in case someone is expecting a CreationException or something
+                final Throwable cause = ce.getCause() != null ? ce.getCause() : ce;
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else if (cause instanceof Error) {
+                    throw (Error) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+        return value;
     }
 
     /**
      * Destroy all the instances in this scope
      */
     public void destroy() {
         // TODO We really should ensure no more instances can be added during or after this
-        instances.values().stream().forEach(Instance::destroy);
+        instances.values().forEach((final CompletableFuture<Instance> future) -> {
+            if (future.isDone() && !future.isCompletedExceptionally()) {
```

Comment:
```suggestion
            if (!future.isCompletedExceptionally()) {
```

this may be more correct...